sys/arm64/arm64/locore.S

```diff
@@ -182,38 +322,54 @@ virtdone:
 	/* Make the page table base a virtual address */
 	sub	x26, x26, x29
 	sub	x24, x24, x29
+#ifdef __CHERI_PURE_CAPABILITY__
+	/* Convert modulep into a capability
+	 * XXX-AM: can we set bounds on this?
+	 * TODO: clear perms
+	 */
+	cvtd	c19, x19
+
+	/* Rebuild L1PT and L0PT capabilities */
+	mov	x6, #PAGE_SIZE
+	ldr	x7, =.Lpagetable_clrperms
+	cvtd	c24, x24
+	scbnds	c24, c24, x6
+	clrperm	c24, c24, x7
+	cvtd	c26, x26
+	scbnds	c26, c26, x6
+	clrperm	c26, c26, x7
+#endif
 
-	sub	sp, sp, #BOOTPARAMS_SIZE
-	mov	x0, sp
-
-	/* Degate the delda so it is VA -> PA */
+	/* Negate the delta so it is VA -> PA */
 	neg	x29, x29
 
-	str	x19, [x0, #BP_MODULEP]
-	str	x26, [x0, #BP_KERN_L1PT]
-	str	x29, [x0, #BP_KERN_DELTA]
-	adr	x25, initstack
-	str	x25, [x0, #BP_KERN_STACK]
-	str	x24, [x0, #BP_KERN_L0PT]
-	str	x23, [x0, #BP_BOOT_EL]
-	str	x27, [x0, 40]	/* kern_ttbr0 */
+	sub	PTRN(sp), PTRN(sp), #BOOTPARAMS_SIZE
+	mov	PTR(0), PTRN(sp)
+
+	str	PTR(19), [PTR(0), #BP_MODULEP]
+	str	PTR(26), [PTR(0), #BP_KERN_L1PT]
+	str	x29, [PTR(0), #BP_KERN_DELTA]
+	str	PTR(25), [PTR(0), #BP_KERN_STACK]
+	str	PTR(24), [PTR(0), #BP_KERN_L0PT]
+	str	x23, [PTR(0), #BP_BOOT_EL]
+	str	x27, [PTR(0), #BP_KERN_TTBR0]
 
 	/* trace back starts here */
 	mov	fp, #0
 	/* Branch to C code */
+	clear_capregs
+
```

Comment:
Spacing is a bit weird, and not sure if clear_capregs should be before the comment too

Reply:
I would keep it below the comment because I would argue that clearing the capability registers from everything left over from assembly is part of entering the C part of the boot.